In [1]:
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
titles = pd.read_csv('data/titles.csv')
titles.head()

title  year
0    Tasveer Mere Sanam  1996
1              Only You  1994
2  El pueblo del terror  1970
3               Machine  2007
4                  MARy  2008

### How many movies are listed in the titles dataframe?

In [3]:
titles.shape

(244914, 2)

Thus there are 244914 movies listed in the dataframe with duplicates

### What are the earliest two films listed in the titles dataframe?

In [4]:
titles.sort_values(by = 'year').head(2)

title  year
193222          Miss Jerry  1894
51388   The Startled Lover  1898

Miss Jerry and the Startled Lover are the two eariest films

### How many movies have the title "Hamlet"?

In [5]:
movies = titles.pivot_table(index=['title'], aggfunc='size')
movies['Hamlet']

20

There are 20 Movies titled Hamlet

### How many movies are titled "North by Northwest"?

In [6]:
movies['North by Northwest']

1

There are 1 movie titled North by Northwest

### When was the first movie titled "Hamlet" made?

In [7]:
sort_dates = titles.sort_values(by = 'year')
sort_dates = sort_dates[sort_dates['title'] == 'Hamlet']
sort_dates[sort_dates.year == sort_dates.year.min()]

title  year
52521  Hamlet  1910

### List all of the "Treasure Island" movies from earliest to most recent.

In [8]:
titles[titles['title'] == 'Treasure Island'].sort_values(by = 'year')

title  year
55686   Treasure Island  1918
196514  Treasure Island  1920
141822  Treasure Island  1934
77354   Treasure Island  1950
65051   Treasure Island  1972
37238   Treasure Island  1973
173051  Treasure Island  1985
27076   Treasure Island  1999

### How many movies were made in the year 1950?

In [9]:
year_count = titles.pivot_table(index=['year'], aggfunc='size')
year_count[1950]

1113

### How many movies were made in the year 1960?

In [10]:
year_count[1960]

1549

### How many movies were made from 1950 through 1959?

In [11]:
movies_bw = titles[(titles['year'] >= 1950) & (titles['year'] <= 1959)]
movies_bw['title'].value_counts().sum()

12934

### In what years has a movie titled "Batman" been released?

In [12]:
titles[titles['title'] == 'Batman'].head()

title  year
39349   Batman  1943
240896  Batman  1989

In [13]:
cast = pd.read_csv('data/cast.csv')
cast.head()

title  year      name   type                character     n
0        Closet Monster  2015  Buffy #1  actor                  Buffy 4  31.0
1       Suuri illusioni  1985    Homo $  actor                   Guests  22.0
2   Battle of the Sexes  2017   $hutter  actor          Bobby Riggs Fan  10.0
3  Secret in Their Eyes  2015   $hutter  actor          2002 Dodger Fan   NaN
4            Steve Jobs  2015   $hutter  actor  1988 Opera House Patron   NaN

### How many roles were there in the movie "Inception"?

In [14]:
temp = cast[cast['title'] == 'Inception']
temp['name'].value_counts().sum()

78

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [15]:
temp = cast[(cast['title'] == 'Inception') & (cast.n.isna() == True)]
temp['name'].value_counts().sum()

27

### But how many roles in the movie "Inception" did receive an "n" value?

In [16]:
temp = cast[(cast['title'] == 'Inception') & (cast.n.isna() == False)]
temp['name'].value_counts().sum()

51

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [17]:
temp = cast[cast['title'] == 'North by Northwest'].sort_values(by = 'n')
temp['name'].head(30)

871426               Cary Grant
3518333         Eva Marie Saint
1457165             James Mason
3157232     Jessie Royce Landis
356707           Leo G. Carroll
3048272    Josephine Hutchinson
1695173             Philip Ober
1275337           Martin Landau
2447081           Adam Williams
1811387            Edward Platt
665768        Robert Ellenstein
2295714            Les Tremayne
464736          Philip Coolidge
1508344           Patrick McVey
206471             Edward Binns
1384780               Ken Lynch
12668             Stanley Adams
29122                Andy Albin
56389              Don Anderson
56610           Ernest Anderson
95950         Malcolm Atterbury
101764                Tol Avery
113558               Sam Bagley
119104              Frank Baker
124956                Alex Ball
143362            Baynes Barron
157984            Brandon Beach
183104            John Beradino
231150          Stephen Bolster
357237         Steve Carruthers
Name: name, dtype: object

In [18]:
temp['name'].value_counts().sum()

116

### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [19]:
temp = cast[(cast['title'] == 'Sleuth') & (cast['year'] == 1972)].sort_values(by = 'n')
temp['name']

1705394      Laurence Olivier
326773          Michael Caine
373876         Alec Cawthorne
1465483    John (II) Matthews
2724553    Eve (III) Channing
1448690          Teddy Martin
Name: name, dtype: object

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [20]:
temp = cast[(cast['title'] == 'Sleuth') & (cast['year'] == 2007)].sort_values(by = 'n')
temp['name']

326774           Michael Caine
1293634               Jude Law
1805907          Harold Pinter
260242         Kenneth Branagh
373877     Alec (II) Cawthorne
2724552      Eve (II) Channing
3370424      Carmel O'Sullivan
Name: name, dtype: object

### How many roles were credited in the silent 1921 version of Hamlet?

In [21]:
temp = cast[(cast['title'] == 'Hamlet') & (cast['year'] == 1921)]
temp['name'].value_counts().sum()

9

### How many roles were credited in Branagh’s 1996 Hamlet?

In [22]:
temp = cast[(cast['title'] == 'Hamlet') & (cast['year'] == 1996)]
(temp['name'] == 'Kenneth Branagh').value_counts()

False    53
True      1
Name: name, dtype: int64

### How many "Hamlet" roles have been listed in all film credits through history?

In [23]:
temp = cast[cast['character'] == 'Hamlet']
temp['character'].value_counts().sum()

96

### How many people have played an "Ophelia"?

In [24]:
temp = cast[cast['character'] == 'Ophelia']
temp['character'].value_counts().sum()

117

### How many people have played a role called "The Dude"?

In [25]:
temp = cast[cast['character'] == 'The Dude']
temp['character'].value_counts().sum()

19

### How many people have played a role called "The Stranger"?

In [26]:
temp = cast[cast['character'] == 'The Stranger']
temp['character'].value_counts().sum()

212

### How many roles has Sidney Poitier played throughout his career?

In [27]:
temp = cast[cast['name'] == 'Sidney Poitier']
temp['character'].value_counts().sum()

43

### How many roles has Judi Dench played?

In [28]:
temp = cast[cast['name'] == 'Judi Dench']
temp['character'].value_counts().sum()

55

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [29]:
temp = cast[(cast['name'] == 'Cary Grant') & (cast['character'] != 'Lead') & (cast['n'] == 2) & ((cast['year'] >= 1940) & (cast['year'] < 1950))].sort_values(by = 'year')
temp['character']

871423     Nick Arden
871433    Roger Adams
Name: character, dtype: object

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [30]:
temp = cast[(cast['name'] == 'Cary Grant') & (cast['n'] == 1) &((cast['year'] >= 1940) & (cast['year'] < 1950))].sort_values(by = 'year')
temp['character']

871448                             Matt Howard
871407                            Walter Burns
871450                      C. K. Dexter Haven
871438                        Johnnie Aysgarth
871452                            Leopold Dilg
871429                   Patrick 'Pat' O'Toole
871398                           Capt. Cassidy
871422    Joe Adams -posing as Joe Bascopolous
871430                             Jerry Flynn
871390                       Mortimer Brewster
871425                              Ernie Mott
871424                             Cole Porter
871427                                  Devlin
871444                             Dick Nugent
871445                                  Dudley
871421                           Jim Blandings
871402                       Dr. Madison Brown
871411                   Captain Henri Rochard
Name: character, dtype: object

### How many roles were available for actors in the 1950s?

In [31]:
temp = cast[(cast['type'] == 'actor') & ((cast['year'] >= 1950) & (cast['year'] < 1960))]
temp['character'].value_counts().sum()

157738

### How many roles were available for actresses in the 1950s?

In [32]:
temp = cast[(cast['type'] == 'actress') & ((cast['year'] >= 1950) & (cast['year'] < 1960))]
temp['character'].value_counts().sum()

57733

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [33]:
temp = cast[(cast['n'] == 1) & (cast['year'] == 1980)]
temp['character'].value_counts().sum()

1183

### How many non-leading roles were available through from the beginning of film history through 1980?

In [34]:
temp = cast[(cast['n'] != 1) & (cast['year'] == 1980)]
temp['character'].value_counts().sum()

22289

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [35]:
temp = cast[(cast['n'].isna() == True) & (cast['year'] == 1980)]
temp['character'].value_counts().sum()

4466